drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,511 @@
+
+	HAL_DCMI_IRQHandler(&data->hdcmi);
+}
+
+static void dmci_dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	DMA_HandleTypeDef *hdma = arg;
+
+	ARG_UNUSED(dev);
+
+	if (status < 0) {
+		LOG_ERR("DMA callback error with channel %d.", channel);
+	}
+
+	HAL_DMA_IRQHandler(hdma);
+}
+
+void HAL_DMA_ErrorCallback(DMA_HandleTypeDef *hdma)
+{
+	LOG_WRN("%s", __func__);
+}
+
+static int stm32_dma_init(const struct device *dev)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+	int ret;
+
+	/* Check if the DMA device is ready */
+	if ((data->dma.dma_dev != NULL) &&
+		!device_is_ready(data->dma.dma_dev)) {
+		LOG_ERR("%s DMA device not ready", data->dma.dma_dev->name);
+		return -ENODEV;
+	}
+
+	/*
+	 * DMA configuration
+	 * Due to use of QSPI HAL API in current driver,
+	 * both HAL and Zephyr DMA drivers should be configured.
+	 * The required configuration for Zephyr DMA driver should only provide
+	 * the minimum information to inform the DMA slot will be in used and
+	 * how to route callbacks.
+	 */
+	struct dma_config dma_cfg = data->dma.cfg;
+	static DMA_HandleTypeDef hdma;
+
+	/* Proceed to the minimum Zephyr DMA driver init */
+	dma_cfg.user_data = &hdma;
+	/* HACK: This field is used to inform driver that it is overridden */
+	dma_cfg.linked_channel = STM32_DMA_HAL_OVERRIDE;
+	/* Because of the STREAM OFFSET, the DMA channel given here is from 1 - 8 */
+	ret = dma_config(data->dma.dma_dev,
+			 (data->dma.channel + STM32_DMA_STREAM_OFFSET), &dma_cfg);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure DMA channel %d",
+			data->dma.channel + STM32_DMA_STREAM_OFFSET);
+		return ret;
+	}
+
+	/*** Configure the DMA ***/
+	/* Set the parameters to be configured */
+	hdma.Init.Request		= DMA_REQUEST_DCMI;
+	hdma.Init.Direction		= DMA_PERIPH_TO_MEMORY;
+	hdma.Init.PeriphInc		= DMA_PINC_DISABLE;
+	hdma.Init.MemInc		= DMA_MINC_ENABLE;
+	hdma.Init.PeriphDataAlignment	= DMA_PDATAALIGN_WORD;
+	hdma.Init.MemDataAlignment	= DMA_MDATAALIGN_WORD;
+	hdma.Init.Mode			= DMA_CIRCULAR;
+	hdma.Init.Priority		= DMA_PRIORITY_HIGH;
+	hdma.Init.FIFOMode		= DMA_FIFOMODE_DISABLE;
+
+	hdma.Instance = __LL_DMA_GET_STREAM_INSTANCE(data->dma.reg,
+						data->dma.channel);
+
+	/* Initialize DMA HAL */
+	__HAL_LINKDMA(&data->hdcmi, DMA_Handle, hdma);
+
+	if (HAL_DMA_Init(&hdma) != HAL_OK) {
+		LOG_ERR("DCMI DMA Init failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int stm32_dcmi_enable_clock(const struct device *dev)
+{
+	const struct video_stm32_dcmi_config *config = dev->config;
+	const struct device *dcmi_clock = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	int err;
+
+	if (!device_is_ready(dcmi_clock)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Turn on DCMI peripheral clock */
+	err = clock_control_on(dcmi_clock, (clock_control_subsys_t *) &config->pclken);
+	if (err < 0) {
+		LOG_ERR("Failed to enable DCMI clock");
+		LOG_ERR("Error code %d: %s", err, strerror(-err));
```

Comment:
I would avoid using `strerror` in driver code, I think that when you do that you end up compiling in the error string table which may be expensive, just log the error code